[CMake] Use imported targets in the GLib find-module

aperezdc opened the pull request:
#### 68a9c2d8e124d58896ee02cff1eb2fd87feb0ae6
<pre>
[CMake] Use imported targets in the GLib find-module
<a href="https://bugs.webkit.org/show_bug.cgi?id=299597">https://bugs.webkit.org/show_bug.cgi?id=299597</a>

Reviewed by Patrick Griffis.

Rewrite the GLib CMake find-module to use imported targets, simplifying
the CMake build system thanks to them. While at it rename it as
FindGLib.cmake to make it more idiomatic.

The new find-module defines a GLib::GLib imported target for the main
library, and sub-targets for the different components (GLib::Thread,
GLib::Object, etc.) which encode the dependencies between them. The
latter means that linking e.g. GLib::Gio into a target does not need
indicating also GLib::Object and GLib::GLib because CMake will
automatically pull the compile flags and libraries from the dependent
targets.

* Source/JavaScriptCore/GLib.cmake:
* Source/JavaScriptCore/PlatformGTK.cmake:
* Source/JavaScriptCore/PlatformJSCOnly.cmake:
* Source/JavaScriptCore/PlatformWPE.cmake:
* Source/JavaScriptCore/inspector/remote/GLib.cmake:
* Source/JavaScriptCore/shell/PlatformGTK.cmake: Removed.
* Source/JavaScriptCore/shell/PlatformWPE.cmake: Removed.
* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WebCore/PAL/pal/PlatformWPE.cmake:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/drm/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/cmake/FindGLIB.cmake: Removed.
* Source/cmake/FindGLib.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/WebKitTestRunner/PlatformWPE.cmake:
* Tools/wpe/backends/PlatformWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/300628@main">https://commits.webkit.org/300628@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f70a561731b5a62550968b424183a17f3629ed3a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129939 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75345 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125109 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43670 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51541 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93676 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62149 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126185 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34806 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110276 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74304 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33772 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28432 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73455 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115391 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104520 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28658 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132657 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121763 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50182 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38193 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102165 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50558 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106497 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102022 "Found 1 new API test failure: WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25949 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47380 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25599 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46956 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50037 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55798 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/152118 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49507 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38872 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52858 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/FindGLib.cmake:
```diff
@@ -0,0 +1,158 @@
+# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#[=======================================================================[.rst:
+#]=======================================================================]
```
Ouch, I forgot to add the find-module documentation here. Expect an update with that 😉

TingPing commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -8,7 +8,7 @@ SET_PROJECT_VERSION(2 51 0)
 
 # Update Source/WTF/wtf/Platform.h to match required GLib versions.
-find_package(GLIB 2.70.0 REQUIRED COMPONENTS gio gio-unix gobject gthread gmodule)
+find_package(GLib 2.70.0 REQUIRED COMPONENTS Gio GioUnix Object Thread Module)
```
Shouldn't `GioUnix` imply `Object` and `Gio`. 

Same elsewhere.

aperezdc replied:
Good point! I changed this before having the idea of adding the support to declare the dependencies among modules. We should be able to remove `Object` and `Gio` from the list here, I'll try.

TingPing reviewed (approved):
Nice cleanup

mcatanzaro commented on Source/JavaScriptCore/GLib.cmake:
```diff
@@ -25,6 +25,8 @@ list(APPEND JavaScriptCore_SOURCES
     API/glib/JSCWrapperMap.cpp
 )
 
+list(APPEND JavaScriptCore_PRIVATE_LIBRARIES GLib::GLib)
```
I suspect this works only because something somewhere else is using `GLib::Gio`.

Also, it might actually be useful to _not_ use `PRIVATE` here, so it gets "inherited" into targets that depend on JavaScriptCore.

aperezdc replied:
Indeed, removing `PRIVATE` here makes sense because code using the JSC GLib API needs to inherit the build flags.

aperezdc replied:
Done!

mcatanzaro commented on Source/WTF/wtf/PlatformWPE.cmake:
```diff
@@ -66,9 +66,7 @@ list(APPEND WTF_PUBLIC_HEADERS
-    ${GLIB_GIO_LIBRARIES}
-    ${GLIB_GOBJECT_LIBRARIES}
-    ${GLIB_LIBRARIES}
+    GLib::Gio
```
Since this is public, shouldn't it already get used by everything that depends on WTF, which should be... just about everything? In theory, we should be able to specify it just here and not need to do so in other places?

aperezdc replied:
In theory, yes. I'll give it a try and in general check around where there might be other places where we can simplify a bit more.

mcatanzaro reviewed (commented):
Nice improvement.

In theory, we should be doing this for all our find modules. Too bad we still need find modules at all. This feels so primitive compared to pkg-config. :(

aperezdc committed:
[CMake] Use imported targets in the GLib find-module
https://bugs.webkit.org/show_bug.cgi?id=299597

Reviewed by Patrick Griffis.

Rewrite the GLib CMake find-module to use imported targets, simplifying
the CMake build system thanks to them. While at it rename it as
FindGLib.cmake to make it more idiomatic.

The new find-module defines a GLib::GLib imported target for the main
library, and sub-targets for the different components (GLib::Thread,
GLib::Object, etc.) which encode the dependencies between them. The
latter means that linking e.g. GLib::Gio into a target does not need
indicating also GLib::Object and GLib::GLib because CMake will
automatically pull the compile flags and libraries from the dependent
targets.

* Source/JavaScriptCore/GLib.cmake:
* Source/JavaScriptCore/PlatformGTK.cmake:
* Source/JavaScriptCore/PlatformJSCOnly.cmake:
* Source/JavaScriptCore/PlatformWPE.cmake:
* Source/JavaScriptCore/inspector/remote/GLib.cmake:
* Source/JavaScriptCore/shell/PlatformGTK.cmake: Removed.
* Source/JavaScriptCore/shell/PlatformWPE.cmake: Removed.
* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WebCore/PAL/pal/PlatformWPE.cmake:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/drm/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/cmake/FindGLIB.cmake: Removed.
* Source/cmake/FindGLib.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsJSCOnly.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/WebKitTestRunner/PlatformWPE.cmake:
* Tools/wpe/backends/PlatformWPE.cmake:

Canonical link: https://commits.webkit.org/300628@main
